src/mds/Locker.cc

```diff
@@ -4680,7 +4716,41 @@ void Locker::scatter_tick()
     if (now - lock->get_update_stamp() < g_conf()->mds_scatter_nudge_interval)
       break;
     updated_scatterlocks.pop_front();
-    scatter_nudge(lock, 0);
+    if (lock->get_type() == CEPH_LOCK_INEST) {
+      set<MDRequestRef> mdrs;
+      for (auto mdr : mdrs_inflight){
+    	if (mdr->in[0]->is_projected_ancestor_of(static_cast<CInode *>(lock->get_parent())) &&
+	    ((!mdr->in[0]->get_rstat_dirty_from().is_zero()
+	      && mdr->in[0]->get_rstat_dirty_from() <= mdr->rstats_propagate_time)
+	     || mdr->in[0]->get_dirfrags_rstat_dirty_from() <= mdr->rstats_propagate_time)
+	    && !mdr->locks_inflight.count(lock)) {
```

Comment:
I'd like to propagate all nestlocks dirtied before the specified time, instead of doing is_projected_ancestor_of check

Besides, I don't want Locker:: mdrs_inflight and MDRequest::locks_inflight. What I like is  roughly sorted list for dirty nestlocks, like:

map<utime_t, xlist<ScatterLock*> > updated_nestlocks.



The map only have one item If no one  forcedly propagate rstat.  when forcedly propagating rstat, we add new list 
(key is propagate_time) to the map.  mark_updated_scatterlock() adds dirty nestlocks to specific list according to rstat_dirty_from

Reply:
Ok, I'll correct that:-)

Reply:
@ukernel Hi, it seems that if we remove is_projected_ancestor_of check, dirtied rstats would be propagated to the root, which means users do "setfattr -n ceph.rstats.propagate -v 1 /a/b" for a specific directory "/a/b" while, finally, they propagate the whole file system. This seems pointless and would make the propagation run for a long time. Do we really need to do this? Thanks:-)

Reply:
even do is_projected_ancestor_of check, I'd like to do it in mark_updated_scatterlock(). Data structure looks like. 

map<pair< utime_t, CInode*>,  xlist<ScatterLock*> updated_nestlocks.

For now, I'd like to have a simple and clean implementation.  If you want to do it, do it in incremental patch.